tls/s2n_server_hello.c

```diff
@@ -36,6 +37,54 @@
 /* From RFC5246 7.4.1.2. */
 #define S2N_TLS_COMPRESSION_METHOD_NULL 0
 
+/* From RFC8446 4.1.3. */
+#define S2N_DOWNGRADE_PROTECTION_SIZE   8
+const uint8_t tls12_downgrade_protection_bytes[] = {
+    0x44, 0x4F, 0x57, 0x4E, 0x47, 0x52, 0x44, 0x01
+};
+
+const uint8_t tls11_downgrade_protection_bytes[] = {
+    0x44, 0x4F, 0x57, 0x4E, 0x47, 0x52, 0x44, 0x00
+};
+
+static int s2n_detect_downgrade(struct s2n_connection *conn) {
+    notnull_check(conn);
+    uint8_t *downgrade_protection = &conn->secure.server_random[S2N_TLS_RANDOM_DATA_LEN - S2N_DOWNGRADE_PROTECTION_SIZE];
+
+    /* Detect downgrade attacks according to RFC 8446 section 4.1.3 */
+    if (s2n_is_tls13_enabled() && conn->server_protocol_version == S2N_TLS12) {
+        /* If the server supports TLS1.3, the random data will be set according to RFC sect 4.1.3 */
+        if (memcmp(downgrade_protection, tls12_downgrade_protection_bytes, S2N_DOWNGRADE_PROTECTION_SIZE) == 0) {
```

Comment:
consider s2n_constant_time_equals() instead?

Reply:
Ah yes, I can certainly do that.

Reply:
Updated in next push.